### ISSUE TITLE
DNSSEC records are stripped out even when DNSSEC supported: yes

### ISSUE THREAD

**Issue by pemensik** [systemd/systemd#17272]
**systemd version the issue has been seen with**
 > systemd-246.6-3.fc34.x86_64

**Used distribution**
 > Fedora rawhide

**Linux kernel version used** (`uname -a`)
 > 5.7.11-200.fc32.x86_64

**CPU architecture issue was seen on**
 > x86_64

**Expected behaviour you didn't see**
 > RRSIG records are present on every query with DO bit, where upstream resolver provided them

**Unexpected behaviour you saw**
 > RRSIG records are present on upstream resolver, but local client never get them. While DNSSEC is enabled and reported as supported.

**Steps to reproduce the problem**
```
# rpm -q systemd
systemd-246.6-3.fc34.x86_64


# cat /etc/systemd/network/host.network
[Match]
Name=host0

[Network]
DHCP=yes

[DHCPv4]
UseDNS=yes

[IPV6ACCEPTRA]
UseDNS=yes


# cat /etc/systemd/resolved.conf
#  This file is part of systemd.
#
#  systemd is free software; you can redistribute it and/or modify it
#  under the terms of the GNU Lesser General Public License as published by
#  the Free Software Foundation; either version 2.1 of the License, or
#  (at your option) any later version.
#
# Entries in this file show the compile time defaults.
# You can change settings by editing this file.
# Defaults can be restored by simply deleting this file.
#
# See resolved.conf(5) for details

[Resolve]
#DNS=
#DNS=192.168.88.1
#FallbackDNS=1.1.1.1 1.0.0.1 2606:4700:4700::1111 2606:4700:4700::1001 # Cloudflare
#FallbackDNS=8.8.8.8 8.8.4.4 2001:4860:4860::8888 2001:4860:4860::8844 # Google
#FallbackDNS=9.9.9.9 2620:fe::fe # Quad9
FallbackDNS=
#Domains=
#LLMNR=yes
#MulticastDNS=yes
DNSSEC=yes
#DNSSEC=yes
#DNSOverTLS=no
#Cache=yes
#DNSStubListener=udp


# systemctl restart systemd-networkd systemd-resolved


# resolvectl
Global
       LLMNR setting: resolve
MulticastDNS setting: no     
  DNSOverTLS setting: no     
      DNSSEC setting: yes    
    DNSSEC supported: yes    

Link 2 (host0)
      Current Scopes: DNS LLMNR/IPv4 LLMNR/IPv6
DefaultRoute setting: yes                      
       LLMNR setting: yes                      
MulticastDNS setting: no                       
  DNSOverTLS setting: no                       
      DNSSEC setting: yes                      
    DNSSEC supported: yes                      
  Current DNS Server: 192.168.122.1            
         DNS Servers: 192.168.122.1            


# dig +dnssec @127.0.0.53

; <<>> DiG 9.16.7-RedHat-9.16.7-1.fc34 <<>> +dnssec @127.0.0.53
; (1 server found)
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 34535
;; flags: qr rd ra; QUERY: 1, ANSWER: 13, AUTHORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags: do; udp: 65494
; OPT=5: 05 07 08 0a 0d 0e 0f (".......")
; OPT=6: 01 02 04 ("...")
; OPT=7: 01 (".")
;; QUESTION SECTION:
;.				IN	NS

;; ANSWER SECTION:
.			509999	IN	NS	b.root-servers.net.
.			509999	IN	NS	c.root-servers.net.
.			509999	IN	NS	d.root-servers.net.
.			509999	IN	NS	e.root-servers.net.
.			509999	IN	NS	f.root-servers.net.
.			509999	IN	NS	g.root-servers.net.
.			509999	IN	NS	h.root-servers.net.
.			509999	IN	NS	i.root-servers.net.
.			509999	IN	NS	j.root-servers.net.
.			509999	IN	NS	k.root-servers.net.
.			509999	IN	NS	l.root-servers.net.
.			509999	IN	NS	m.root-servers.net.
.			509999	IN	NS	a.root-servers.net.

;; Query time: 0 msec
;; SERVER: 127.0.0.53#53(127.0.0.53)
;; WHEN: Wed Oct 07 15:32:54 CEST 2020
;; MSG SIZE  rcvd: 262



# dig +dnssec @192.168.122.1

; <<>> DiG 9.16.7-RedHat-9.16.7-1.fc34 <<>> +dnssec @192.168.122.1
; (1 server found)
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 2462
;; flags: qr rd ra; QUERY: 1, ANSWER: 14, AUTHORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags: do; udp: 4096
;; QUESTION SECTION:
;.				IN	NS

;; ANSWER SECTION:
.			509999	IN	NS	d.root-servers.net.
.			509999	IN	NS	e.root-servers.net.
.			509999	IN	NS	f.root-servers.net.
.			509999	IN	NS	g.root-servers.net.
.			509999	IN	NS	h.root-servers.net.
.			509999	IN	NS	i.root-servers.net.
.			509999	IN	NS	j.root-servers.net.
.			509999	IN	NS	k.root-servers.net.
.			509999	IN	NS	l.root-servers.net.
.			509999	IN	NS	m.root-servers.net.
.			509999	IN	NS	a.root-servers.net.
.			509999	IN	NS	b.root-servers.net.
.			509999	IN	NS	c.root-servers.net.
.			509999	IN	RRSIG	NS 8 0 518400 20201020050000 20201007040000 26116 . mXJjxTbIzk9maKOXtJCWsD3sJHmnFCTUjiAiGSHGhhXiJFfWgPo5KgdY xNpiT4+1QMx+kUplFcSbCziqBzJWB9R/bjWAQFkYP64LDxah4QuXBlNW wJc/M2ea+SP1PrZsmNuNACeHd05wNP8bMK+LbWEaBqo+bID+J7Xe4+DJ f7oxfKIqeDwNCdODmsTseFgq1IbBSlw1ppSBR5190552Dta1Ayb4LrtX 3Unfy0gy28hL0AF3ownnAI51Me69OUjaR7E92cJ5/ttjbuZYY4G/6G7n eNWhJ5CGlpwryEtSPBTa9SbiDo2KjG9rYRbSePSCR1tKuZ/SJv81YM6v uCTJ9w==

;; Query time: 0 msec
;; SERVER: 192.168.122.1#53(192.168.122.1)
;; WHEN: Wed Oct 07 15:32:54 CEST 2020
;; MSG SIZE  rcvd: 525



# resolvectl
Global
       LLMNR setting: resolve
MulticastDNS setting: no     
  DNSOverTLS setting: no     
      DNSSEC setting: yes    
    DNSSEC supported: yes    

Link 2 (host0)
      Current Scopes: DNS LLMNR/IPv4 LLMNR/IPv6
DefaultRoute setting: yes                      
       LLMNR setting: yes                      
MulticastDNS setting: no                       
  DNSOverTLS setting: no                       
      DNSSEC setting: yes                      
    DNSSEC supported: yes                      
  Current DNS Server: 192.168.122.1            
         DNS Servers: 192.168.122.1            
```

**Comment by pemensik** [comment]
This extends issue #4621 and was reported in issue #17223. RRSIG is stripped even for DNSSEC compatible client, when systemd-resolved claims DNSSEC is supported. But without RRSIG passed further, no DNSSEC capable client cannot validate a single thing, not even root nameservers.

**Comment by poettering** [comment]
systemd-resolved is primarily a DNS client, not a server. Hence, whenever we talk about DNSSEC support this is really about its own validation logic: i.e. whether it can do or does or shall do DNSSEC validation on its upstream-facing traffic.

The local DNS stub (i.e. app-facing side) only has minimal DNSSEC support, to the point glibc uses it pretty much. There never was the need to support it to more than that since there are barely any clients that actually care about DNSSEC info. Outside of DNSSEC testing tools I still don't know any btw. 

And yes, I am fully aware that there's now a sizeable group of people who want to talk DNSSEC (both with and without CD flag) to systemd-resolved, and we can add that, but please understand where we are coming from. 

Please do not confuse systemd-resolved with a DNS server, it's not one, we do not strive to accept non-local traffic, and implement a full server.

**Comment by poettering** [comment]
Closing this as duplicate of #17223. Key of the issue is that we do not provide RRSIG fields to local DNS stub clients.

**Comment by pemensik** [comment]
Those are different issues however. This issue was about unability to *receive* RRSIG from server. #17223 is about inability to send RRSIG request.

I am completely okay, if systemd-resolved does not act as DNS server. No problem. It can use just nss_resolve to deliver queries for getent hosts fedoraproject.org queries. No problem, this is about everything glibc DNS stub implements.

But as long as it DNS server installed on every machine, it has to talk DNS protocol or use alternative port. Note, nobody needs DNSSEC on MDNS port 5353. But it uses 53, which is used also for DNSSEC.

/etc/resolv.conf contains **nameserver** records. That means DNS server records. If you don't want to implement it, just don't push resolved into this file. As simple as that.

CD flag is mandatory for any debugging of DNSSEC validation failures. If you implement validation, CD must be implemented as well.

DNSSEC can be used also on workstation. ssh is able to use it, but its support is [not yet enabled](https://src.fedoraproject.org/rpms/openssh/pull-request/12). Your disapproval of DNSSEC usage is known, but so far the biggest obstacle for using DNSSEC on desktop is systemd-resolved.

**Comment by mcatanzaro** [comment]
> Closing this as duplicate of #17223. Key of the issue is that we do not provide RRSIG fields to local DNS stub clients.

Looking over both issues, I agree this issue is not a duplicate of #17223, and that #17223 looks much less important.

I think maybe it is a duplicate of #4621, though? Petr, do you agree?

**Comment by pemensik** [comment]
@mcatanzaro Yes, I were a bit mislead by DO/CD bits mentioned. This is no really about CD flag in any way, it is just about DO flag and records not passed to clients. In that sense, I would agree original Ubuntu bug is about the same issue I described.

**Comment by mcatanzaro** [comment]
OK, in that case, we can leave this closed, and continue discussion in #4621. ETOOMANYBUGS otherwise. ;)

**Comment by vdukhovni** [comment]
> systemd-resolved is primarily a DNS client, not a server. Hence, whenever we talk about DNSSEC support this is really about its own validation logic: i.e. whether it can do or does or shall do DNSSEC validation on its upstream-facing traffic.
> 

Well, except from the perspective of applications using the stub resolver APIs in libc, res_nsearch() and the like, systemd-resolver is unquestionably a DNS server, specifically an iterative resolver, no different than a local unbound instance, a corporate DNS cache, or public DNS servers like 8.8.8.8.  When 127.0.0.53 is listed in /etc/resolv.conf, systemd-resolved is the DNS server used by the actual stub resolver code in libc.

> The local DNS stub (i.e. app-facing side) only has minimal DNSSEC support, to the point glibc uses it pretty much. There never was the need to support it to more than that since there are barely any clients that actually care about DNSSEC info. Outside of DNSSEC testing tools I still don't know any btw.

I am a contributor of Postfix, and the primary maintainer of its TLS features.  Postfix supports DANE, and for that to work it requires a correctly working, validating local resolver to be configured in /etc/resolv.conf. This has to perform validation unconditionally correctly, including denial of existence. The same is also true for Exim. So that's at least two clients that care about DNSSEC.

Also as more HTTPS service location information [moves into DNS](https://datatracker.ietf.org/doc/draft-ietf-dnsop-svcb-https/), having trustworthy DNS becomes more important for than just MTAs.  Claiming that clients don't care about DNSSEC is not a particularly forward-looking position. Yeah, if break it badly enough your clients won't be relying on you for the service, but that's not exactly a sound argument.

> And yes, I am fully aware that there's now a sizeable group of people who want to talk DNSSEC (both with and without CD flag) to systemd-resolved, and we can add that, but please understand where we are coming from.
> 
> Please do not confuse systemd-resolved with a DNS server, it's not one, we do not strive to accept non-local traffic, and implement a full server.

It is absolutely a DNS server, and step one is to come to grips with that.  Otherwise, it should not be listed in /etc/resolv.conf and the libc stub resolver libraries should send their queries to real DNS servers, whether local or (when local validation is not yet desired) remote.
